src/zarr/storage/local.py

```diff
@@ -217,28 +217,26 @@ async def exists(self, key: str) -> bool:
         path = self.root / key
         return await asyncio.to_thread(path.is_file)
 
-    async def list(self) -> AsyncGenerator[str, None]:
+    async def list(self) -> AsyncIterator[str]:
         # docstring inherited
         to_strip = str(self.root) + "/"
-        for p in list(self.root.rglob("*")):
-            if p.is_file():
-                yield str(p).replace(to_strip, "")
+        allfiles = [str(p).replace(to_strip, "") for p in self.root.rglob("*") if p.is_file()]
+        return (onefile for onefile in allfiles)
```

Comment:
we don't want to materialize the full `allfiles` list in memory. Can you use `async for` instead? Here is an example of how you could do this:

```python
from collections.abc import AsyncIterator
from abc import ABC, abstractmethod

class Foo(ABC):
    @abstractmethod
    async def foo(self) -> AsyncIterator[str]:
        ...


class Bar(Foo):

    async def helper(self) -> AsyncIterator[str]:
        raise ValueError("")

    async def foo(self) -> AsyncIterator[str]:
        return (x async for x in await self.helper())
```

This code should work and mypy successfully, and doesn't materialize the results of calling `helper`.

Reply:
I'm afraid I don't have enough experience with `async` and no time to learn it in the short term. I am unable to find a solution that pleases MyPy.